### PR TITLE
Added image save functionality

### LINKED ISSUE #3: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.
- [x] A flash message is displayed indicating the image was created successfully

__Discussion Topic__:
- [ ] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by ProbablyFaiz**
Closes #3.

This PR adds the functionality of saving image models (only contain URLs right now) to the database via a form. The URLs are validated using the built in ``uri`` library and tests are added to ensure these views and validation work correctly.

### CODE REVIEW

**Comment by ProbablyFaiz** [comment]
[Issue 2: Save Image Link](https://trello.com/c/K5LiE1XE/7-issue-2-save-image-link)

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
+class ImagesController < ApplicationController
```
Comment (bhargav265) [inline app/controllers/images_controller.rb:1]: Can you remove the index, show edit and destroy routes in the PR? Only the new and create is the scope on the PR.

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  private
+    # Use callbacks to share common setup or constraints between actions.
+    def set_image
+      @image = Image.find(params[:id])
+    end
+
+    # Only allow a trusted parameter "white list" through.
+    def image_params
+      params.fetch(:image, {})
+    end
```
Comment (bhargav265) [inline app/controllers/images_controller.rb]: You can use `params.permit` here. Its supposed to be more secure.

File: app/helpers/images_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module ImagesHelper
```
Comment (bhargav265) [inline app/helpers/images_helper.rb:1]: Can remove this

File: test/fixtures/images.yml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
```
Comment (bhargav265) [inline test/fixtures/images.yml:2]: Can remove this

File: test/models/image_test.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
```
Comment (bhargav265) [inline test/models/image_test.rb:6]: Can discard these lines

File: db/migrate/20200622005738_create_images.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class CreateImages < ActiveRecord::Migration[5.2]
+  def change
+    create_table :images do |t|
+      t.string :url
```
Comment (bhargav265) [inline db/migrate/20200622005738_create_images.rb:4]: can add null: false

File: app/models/image.rb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+class Image < ApplicationRecord
```
Comment (bhargav265) [inline app/models/image.rb:3]: Validation can be added for the URL

File: app/models/image.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+class Image < ApplicationRecord
+  validate :validate_url
+
+  def validate_url
+    uri = URI.parse(url)
+    raise URI::InvalidURIError unless uri.is_a?(URI::HTTP) && !uri.host.nil?
+  rescue URI::InvalidURIError
+    errors.add(:base, 'The URL you provided was invalid.')
+  end
```
Comment (bhargav265) [inline app/models/image.rb:11]: instead of writing so many conditions we can use a predefined gem `valid_url`. It will handle any corner case possible. And make the code bug free.

Comment (bhargav265) [reply app/models/image.rb:11]: Also can you add a test file?

File: app/views/images/_form.html.erb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+      <ul>
+        <% image.errors.full_messages.each do |message| %>
+          <li><%= message %></li>
+        <% end %>
+      </ul>
+    </div>
+  <% end %>
+
+  <div class="form-field">
```
Comment (bhargav265) [inline app/views/images/_form.html.erb:14]: we prefer to use simple_form gem when rendering forms in APM.

File: config/routes.rb [hunk 10]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'home#index'
+  resources :images, only: %i[new create show]
```
Comment (bhargav265) [inline config/routes.rb:4]: Can you separate out these actions in to 3 commits?

File: app/views/images/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<p id="notice"><%= notice %></p>
+
+<img src="<%= @image.url %>"  alt="<%= @image.url %>">
+<br/>
+<!--<%#= link_to 'Edit', edit_image_path(@image) %> |-->
```
Comment (bhargav265) [inline app/views/images/show.html.erb:5]: Can remove the comment

File: app/assets/stylesheets/application.css [hunk 12]
```diff
@@ -13,3 +13,7 @@
  *= require_tree .
  *= require_self
  */
+
```
Comment (bhargav265) [inline app/assets/stylesheets/application.css:16]: not required for the story but its okay. We are going away from custom CSS as much as we can

File: config/routes.rb (hunk 10 above)
Comment (ProbablyFaiz) [reply config/routes.rb:4]: I’ll split it into two because new/create go hand in hand.

File: app/models/image.rb (hunk 8 above)
Comment (ProbablyFaiz) [reply app/models/image.rb:11]: The validation tests are done in the controller test file.

File: app/models/image.rb [hunk 13]
```diff
@@ -0,0 +1,3 @@
+class Image < ApplicationRecord
+  validates :url, url: true
```
Comment (bhargav265) [inline app/models/image.rb:2]: Can you add unit tests for this?

File: Gemfile [hunk 14]
```diff
@@ -19,6 +19,8 @@ gem 'jquery-rails'
 
 gem 'valid_url'
 
+gem 'simple_form'
```
Comment (bhargav265) [inline Gemfile:22]: This can be in a separate commit

File: Gemfile [hunk 15]
```diff
@@ -17,6 +17,8 @@ gem 'jquery-rails'
 # Use ActiveModel has_secure_password
 # gem 'bcrypt', '~> 3.1.7'
 
+gem 'valid_url'
```
Comment (bhargav265) [inline Gemfile:20]: This can be in a separate commit

File: test/fixtures/images.yml [hunk 16]
```diff
@@ -0,0 +1,8 @@
+valid_image_url:
+  url: https://github.com/ProbablyFaiz.png
+
+wrong_url_type:
+  url: ftp://example.com/test.png
+
+not_url:
+  url: hello
```
Comment (bhargav265) [inline test/fixtures/images.yml:8]: You dont need to have. fixtures. To test you can create a temp image in the test itself.

File: app/controllers/images_controller.rb [hunk 17]
```diff
@@ -17,6 +22,11 @@ def create
 
   private
 
+  # Use callbacks to share common setup or constraints between actions.
+  def set_image
+    @image = Image.find(params[:id])
```
Comment (bhargav265) [inline app/controllers/images_controller.rb:27]: can you permit this id here?

Also can you rescue if the id isnt there?

Comment (ProbablyFaiz) [reply app/controllers/images_controller.rb:27]: The id is definitely permitted if I understand your question.

> Also can you rescue if the id isnt there?

Default Rails behavior will handle an ActiveRecord::RecordNotFound exception by responding with a 404, which is the behavior you want in 99.99% of cases. This should be kept the way it is, unless there is a really good reason to depart from Rails convention (this method even comes directly from the scaffold).

File: test/models/image_test.rb [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
+class ImageTest < ActiveSupport::TestCase
+  test 'valid image with valid url' do
+    image = Image.new(url: 'https://github.com/ProbablyFaiz.png')
+    assert image.valid?
+  end
+
+  test 'invalid image when url is wrong type' do
+    image = Image.new(url: 'ftp://example.com/test.png')
+    refute image.valid?
```
Comment (bhargav265) [inline test/models/image_test.rb:11]: You can test the validation error message too here.

File: test/models/image_test.rb [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+  end
+
+  test 'invalid image when url is not url' do
+    image = Image.new(url: 'hello')
+    refute image.valid?
```
Comment (bhargav265) [inline test/models/image_test.rb:16]: Same. Would also suggest writing a test for `nil` url

File: app/views/images/_form.html.erb [hunk 20]
```diff
@@ -0,0 +1,4 @@
+<%= simple_form_for @image do |f| %>
```
Comment (bhargav265) [inline app/views/images/_form.html.erb:1]: This is nice

File: test/controllers/images_controller_test.rb [hunk 21]
```diff
@@ -0,0 +1,27 @@
+require 'test_helper'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+  test 'should get new' do
+    get new_image_url
+    assert_response :success
```
Comment (bhargav265) [inline test/controllers/images_controller_test.rb:6]: can we do an assert_select here too?

File: app/controllers/images_controller.rb (hunk 17 above)
Comment (bhargav265) [reply app/controllers/images_controller.rb:27]: I was suggesting you to do something similar to this 
```
  def image_params
    params.require(:image).permit(:url)
  end	
```

Yes, in APM most errors are rescued and probably the index page should be rendered with an error message. I know that showing the error message is out of the scope, but you could just redirect to the index page if the image doesn't exist.

for example try going to `https://qa.qa.appfolio.com/properties/0903` and see what happens

File: app/controllers/images_controller.rb [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  # POST /images
+  def create
+    @image = Image.new(image_params)
+
+    if @image.save
+      redirect_to @image, notice: 'Image was successfully created.'
+    else
+      render :new
```
Comment (bhargav265) [inline app/controllers/images_controller.rb:19]: can we give a `status :unprocessable_entity` here?

File: test/controllers/images_controller_test.rb [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  test 'should create image for valid url' do
+    assert_difference('Image.count', 1) do
+      post images_url, params: { image: { url: 'https://github.com/ProbablyFaiz.png' } }
+    end
+    assert_redirected_to image_url(Image.last)
```
Comment (bhargav265) [inline test/controllers/images_controller_test.rb:13]: can you assert the notice here?

File: app/controllers/images_controller.rb [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+  def new
+    @image = Image.new
+  end
+
+  # POST /images
+  def create
+    @image = Image.new(image_params)
+
+    if @image.save
+      redirect_to @image, notice: 'Image was successfully created.'
```
Comment (bhargav265) [inline app/controllers/images_controller.rb:17]: This would fail, if we run the app at this commit because the show route is still not there. I would either suggest you put the show commit first and then do this commit. Or redirect to the index page here and change it later to show in the show commit. Whatever works for you. I am fine with either.

File: app/controllers/images_controller.rb (hunk 17 above)
Comment (ProbablyFaiz) [reply app/controllers/images_controller.rb:27]: For the params, ids are permitted by default so adding that wouldn’t do anything.

Good point on redirecting to index, but I’m going to leave it as is here because a 404 is the most predictable behavior IMO

File: app/controllers/images_controller.rb (hunk 24 above)
Comment (ProbablyFaiz) [reply app/controllers/images_controller.rb:17]: True, will do.

File: app/controllers/images_controller.rb (hunk 17 above)
Comment (bhargav265) [reply app/controllers/images_controller.rb:27]: You can still return a `status :not_found` along with a redirect. And it is not a good user experience if we show the default rails error page to the user. This was supposed to be an acceptance criteria but it got removed somehow.

Comment (bhargav265) [reply app/controllers/images_controller.rb:27]: About permitting I was talking about this https://github.com/rails/strong_parameters#permitted-scalar-values
